Restructuring API documentation (#160)
Altinn/altinn-studio-docs: content/teknologi/altinnstudio/altinn-api/app-api/data-element/_index.md

> ---

Reviewer: Could we rename file to match title? -> data-elements

title: Data elements
description: The App API to work with the data elements associated with an instance
toc: true
tags: [api]
weight: 100
---

## Overview

> A Data element consist of two parts; its metadata document and the actual data blob. This API work primarily with the blob while keeping the metadata document updated.

Reviewer: ```suggestion
A data element consist of two parts; its metadata document and the actual data blob. This API work primarily with the blob while keeping the metadata document updated.
```

**basePath**
```http
{org}/{appname}/instances/{instanceOwnerPartyId}/{instanceGuid}/data
```

## Get data 

Endpoint for downloading the data blob.

```http
GET basePath/{dataGuid}

> Accept: application/{xml/json}

Reviewer: does the accept header need some more explanation? It isn't clear to me what I would have to use. Do we use application/json for all attachments ?

> Accept: application/{xml/json}

Reviewer: Asking because I don't know.. Will this accept header allow for the return of attachments? 
Or should this information be linked to the form data section?

Author: The accept header is used when picking a serializer. This isn't done by us in our code, but by the framework. All we do is call the `Ok` or `Created` method and the framework does the rest. For Attachments we use the `File` method, which I believe ignores the Accept header and instead use the Content-Type given to it as a parameter.

```

> The response will depend on the type of blob. There are currently 2 primary types; forms and attachments.

Reviewer: I would like to see the terms here be repeated in the sub section headers.. e.g. forms -> Response with form or form data -> respose with form data.

### Response with form data

The respons will either be a json or xml serialized version of the object model depending on the **Accept** header in the request. The value **application/xml** will result in an XML document and the value **application/json** will result inn a JSON document.

### Response with instance attachment

The respons will be a file stream. Content-Type will be the same as the original value given when the file was uploaded. The same is true for the file name.

```http
Content-Disposition: attachment; filename=cute_cat.png; filename*=UTF-8''cute_cat.png
Content-Length: 16994
Content-Type: image/png
```

## Upload data

> Endpoint for initially uploading a new data element on a specific instance.

Reviewer: ```suggestion
Endpoint for uploading a new data element on a specific instance.
```

```http
POST basePath?dataType={data type name}
```

> The **dataType** parameter is required and should contain the name of one of the data types defined on the application. Data types with an **appLogic** property are linked to a form and will have data validation and calculation rules associated with them. Submitted data will be deserialized from **application/xml** or **application/json** based on the content-type of the request. Data types without an **appLogic** property will be handeled as any random file and streamed directly to storage.

Reviewer: ```suggestion
The **dataType** parameter is required and should reference one of the data types defined on the application. Data types with an **appLogic** property are linked to a form and will have data validation and calculation rules associated with them. Submitted data will be deserialized from **application/xml** or **application/json** based on the content-type of the request. Data types without an **appLogic** property will be handeled as any random file and streamed directly to storage.
```

> The **dataType** parameter is required and should contain the name of one of the data types defined on the application. Data types with an **appLogic** property are linked to a form and will have data validation and calculation rules associated with them. Submitted data will be deserialized from **application/xml** or **application/json** based on the content-type of the request. Data types without an **appLogic** property will be handeled as any random file and streamed directly to storage.

Reviewer: ```suggestion
The **dataType** parameter is required and should contain the name of one of the data types defined on the application. Data types with an **appLogic** property are linked to a form and will have data validation and calculation rules associated with them. Submitted data will be deserialized from **application/xml** or **application/json** based on the content-type of the request. Data types without an **appLogic** property will be handeled as an attachment and streamed directly to storage.
```

> ### Submitting form as application/json

Reviewer: ```suggestion
### Uploading form data as application/json
```

Maybe this would link it closer to the upload topic? Only a personal preference as submitting has other connotations for me.

```json
Content-Type: application/json

{
    "dataFormatProvider": "SERES",
    "dataFormatId": "5703",
    "dataFormatVersion": "34553",
    "Tjenesteeier": null,
    "Kontaktperson": {
        "navn": "Sophie Salt",
        "epost": "1337@altinnstudiotestusers.com",
        "telefonnummer": "90001337"
    },
    "OEnsketBruk": null
}

```

> ### Submitting form as application/xml

Reviewer: ```suggestion
### Uploading form data as application/xml
```

```xml
Content-Type: application/xml

<BliTjenesteeier_M xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" dataFormatProvider="SERES" dataFormatId="5703" dataFormatVersion="34553">
    <Kontaktperson>
        <navn>Sophie Salt</navn>
        <epost>1337@altinnstudiotestusers.com</epost>
        <telefonnummer>90001337</telefonnummer>
    </Kontaktperson>
</BliTjenesteeier_M>
```

> ### Submitting attachment

Reviewer: ```suggestion
### Uploading an attachment
```

Most web api client frameworks will have support for creating a valid file upload request for binary data.

```http
Content-Type: application/pdf
Content-Disposition: attachment; filename=receipt.pdf; filename*=UTF-8''receipt.pdf
Content-Length: 16994

%PDF-1.4
%Óëéá
1 0 obj
...
```

> ### Upload response

Reviewer: Going with upload in the header for the previous subsections might require re-thinking this header as they don't really do together.. Will be a bit confusing.

The endpoint returns the data element metadata document that was created.

```json
{
    "id": "692ee7df-82a9-4bba-b2f2-c8c4dac69aff",
    "instanceGuid": "762011d1-d341-4c0a-8641-d8a104e83d30",
    "dataType": "default",
    "contentType": "application/xml",
    "blobStoragePath": "org/app/762011d1-d341-4c0a-8641-d8a104e83d30/data/692ee7df-82a9-4bba-b2f2-c8c4dac69aff",
    "selfLinks": {
        "apps":   "{appPath}/instances/347829/762011d1-d341-4c0a-8641-d8a104e83d30/data/692ee7df-82a9-4bba-b2f2-c8c4dac69aff",
        "platform": "{storagePath}/instances/347829/762011d1-d341-4c0a-8641-d8a104e83d30/data/692ee7df-82a9-4bba-b2f2-c8c4dac69aff"
    },
    "filename": "default.xml",
    "created": "2019-03-06T15:00:23Z",
    "createdBy": "org23",
    "lastChanged": "2019-03-07T15:00:23Z",
    "lastChangedBy": "org23",
    "size": 20001,
    "locked": false
}
```

## Replace data

> Endpoint for replacing an existing data element with new data. The new data must match the data type of the data element it is replacing.

Reviewer: ```suggestion
Endpoint for replacing the content of an existing data element with new data. The new data must match the data type of the data element it is replacing.
```

```http
PUT basePath/{dataGuid}
```

The endpoint works exactly like the endpoint for uploading a new data element.

## Delete data

Endpoint for deleting an existing data element. It is currently not possible to delete the data for a form this way. 

```http
DELETE basePath/{dataGuid}
```

> ## Validate data

Reviewer: WIP?

Author: Created a separate document instead. Just forgot to remove the header.
